syntax = "proto3";
import "shared/Location.proto";
import "shared/Ack.proto";
import "shared/Worker.proto";
import "google/protobuf/struct.proto";

package mapadroid.mitm_mapper;

service MitmMapper {
  rpc GetLastPossiblyMoved(mapadroid.shared.Worker) returns (LastMoved);
  rpc UpdateLatest(LatestMitmDataEntryUpdateRequest) returns (mapadroid.shared.Ack);
  rpc RequestLatest(LatestMitmDataEntryRequest) returns (LatestMitmDataEntryResponse);
  rpc SetLevel(SetLevelRequest) returns (mapadroid.shared.Ack);
  rpc SetPokestopVisits(SetPokestopVisitsRequest) returns (mapadroid.shared.Ack);
  rpc GetPokestopVisits(mapadroid.shared.Worker) returns (PokestopVisitsResponse);
  rpc GetLevel(mapadroid.shared.Worker) returns (LevelResponse);
  rpc GetInjectionStatus(mapadroid.shared.Worker) returns (InjectionStatus);
  rpc SetInjected(InjectedRequest) returns (mapadroid.shared.Ack);
  rpc GetLastKnownLocation(mapadroid.shared.Worker) returns (LastKnownLocationResponse);
  rpc SetQuestsHeld(SetQuestsHeldRequest) returns (mapadroid.shared.Ack);
  rpc GetQuestsHeld(mapadroid.shared.Worker) returns (GetQuestsHeldResponse);
}

message SetQuestsHeldRequest {
  mapadroid.shared.Worker worker = 1;
  optional QuestsHeld quests_held = 2;
}

message GetQuestsHeldResponse {
  optional QuestsHeld quests_held = 1;
}

message QuestsHeld {
  repeated int32 quest_ids = 1;
}

message SetPokestopVisitsRequest {
  mapadroid.shared.Worker worker = 1;
  int32 pokestop_visits = 2;
}

message SetLevelRequest {
  mapadroid.shared.Worker worker = 1;
  int32 level = 2;
}

message LastKnownLocationResponse {
  optional mapadroid.shared.Location location = 1;
}

message InjectedRequest {
  mapadroid.shared.Worker worker = 1;
  InjectionStatus injected = 2;
}

message InjectionStatus {
  bool is_injected = 1;
}

message LevelResponse {
  int32 level = 1;
}

message PokestopVisitsResponse {
  uint64 stops_visited = 1;
}


message LatestMitmDataEntryUpdateRequest {
  mapadroid.shared.Worker worker = 1;
  string key = 2;
  LatestMitmDataEntry data = 3;
}

message LatestMitmDataEntryResponse {
  optional LatestMitmDataEntry entry = 1;
}

message LatestMitmDataEntryRequest {
  mapadroid.shared.Worker worker = 1;
  string key = 2;
  optional uint64 timestamp_earliest = 3;
}

message LatestMitmDataEntry {
  optional mapadroid.shared.Location location = 1;
  optional uint64 timestamp_received = 2;
  optional uint64 timestamp_of_data_retrieval = 3;
  oneof data {
    google.protobuf.Struct some_dictionary = 4;
    google.protobuf.ListValue some_list = 5;
    bytes raw_message = 6;
  }
}

message LastMoved {
  uint64 timestamp = 1;
}
